### PR TITLE
Plugin installer version 2

### PR DESCRIPTION
**Pull request by denho**
### Submission Review Guidelines:

- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/). 
- Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This PR introduces a new version of plugin installer, allowing to install products purchased on WooCommerce.com much faster and without dependency on WP cron being available on the remote site. 

## Preconditions
- Ensure you have an active subscription to a product on WooCommerce.com. Jot down the product ID.
- Connect the store to your account on WooCommerce.com from **WP admin -> WooCommerce -> Extensions -> My subscriptions.**. Ensure you perform the connection from behalf of admin user of the store.
- Comment out [lines 102-110](https://github.com/woocommerce/woocommerce/blob/2800be9b7bae98b3d43247948671e63a96edb234/plugins/woocommerce/includes/wccom-site/class-wc-wccom-site.php#L102-L110) and [114-122](https://github.com/woocommerce/woocommerce/blob/2800be9b7bae98b3d43247948671e63a96edb234/plugins/woocommerce/includes/wccom-site/class-wc-wccom-site.php#L114-L122) in file `/plugins/woocommerce/includes/wccom-site/class-wc-wccom-site.php` to bypass validation of authentication credentials and allow testing API endpoints through Curl.

## Test a successful installation
1. Send an installation request (ensure to replace strings `<SITE_DOMAIN>` and `<PRODUCT_ID>` with your site domain and product ID you got previously):
```bash
curl --request POST \
  --url 'https://<SITE_DOMAIN>/wp-json/wccom-site/v2/installer?token=XXX&signature=YYY' \
  --header 'Content-Type: application/json' \
  --data '{
	"product-id": <PRODUCT_ID>,
	"run-until-step": "activate_product",
	"idempotency-key": "ABC"
}'
```
2. Wait for a few seconds and confirm you received a response on successful installation, with 200 code and response body similar to the following:
```json
{
	"success": true,
	"state": {
		"product_id": 473617,
		"idempotency_key": "ABC",
		"last_step_name": "activate_product",
		"last_step_status": "completed",
		"last_step_error": null,
		"product_type": "plugin",
		"product_name": "WooCommerce Social Login",
		"already_installed_plugin_info": null,
		"started_seconds_ago": 11
	}
}
```
3. Go to site admin and confirm the plugin has been successfully installed and activated.

## Test an installation error
1. Retry the same request from the previous section.
2. Confirm you received an error response, with 400 code and response body stating the installation has already been run:
```json
{
	"success": false,
	"error_code": "all_installation_steps_run",
	"error_message": "All installation steps have been run",
	"state": {
		"product_id": 473617,
		"idempotency_key": "ABC",
		"last_step_name": "activate_product",
		"last_step_status": "completed",
		"last_step_error": null,
		"product_type": "plugin",
		"product_name": "WooCommerce Social Login",
		"already_installed_plugin_info": {
			"name": "WooCommerce Social Login",
			"version": "2.12.0",
			"active": true
		},
		"started_seconds_ago": 365
	}
}
```

## Test installation state reset
1. Send a request to reset the installation state (ensure to replace strings `<SITE_DOMAIN>` and `<PRODUCT_ID>` with your site domain and product ID you got previously):
```bash
curl --request DELETE \
  --url 'https://<SITE_DOMAIN>/wp-json/wccom-site/v2/installer?token=XXX&signature=YYY' \
  --header 'Content-Type: application/json' \
  --data '{
	"product-id": <PRODUCT_ID>,
	"idempotency-key": "ABC"
}'
```
2. Wait for a few seconds and confirm you received a successful response, with 200 code and response body similar to the following:
```json
{
	"success": true,
	"state": null
}
```

### CODE REVIEW

**Comment by rcstr** [comment]
@denho while testing I noticed that `already_installed_plugin_info` is `null` even after trying to _reinstall_ a few times. I tried with a few products and had the same result

### First installation
```sh
$ curl --request POST   --url 'http://localhost:8888/wp-json/wccom-site/v2/installer?token=XXX&signature=YYY'   --header 'Content-Type: application/json'   --data '{
        "product-id": 5278104,
        "run-until-step": "activate_product",
        "idempotency-key": "ABC"
}'

{
	"success": true,
	"state": {
		"product_id": 5278104,
		"idempotency_key": "ABC",
		"last_step_name": "activate_product",
		"last_step_status": "completed",
		"last_step_error": null,
		"product_type": "plugin",
		"product_name": "WooCommerce Payments",
		"already_installed_plugin_info": null,
		"started_seconds_ago": 3
	}
}
```

### Retry same installation
```sh
$ curl --request POST   --url 'http://localhost:8888/wp-json/wccom-site/v2/installer?token=XXX&signature=YYY'   --header 'Content-Type: application/json'   --data '{
        "product-id": 5278104,
        "run-until-step": "activate_product",
        "idempotency-key": "ABC"
}'

{
	"success": false,
	"error_code": "all_installation_steps_run",
	"error_message": "All installation steps have been run",
	"state": {
		"product_id": 5278104,
		"idempotency_key": "ABC",
		"last_step_name": "activate_product",
		"last_step_status": "completed",
		"last_step_error": null,
		"product_type": "plugin",
		"product_name": "WooCommerce Payments",
		"already_installed_plugin_info": null,
		"started_seconds_ago": 31
	}
}
```

**Comment by thilinah** [comment]
Thanks @denho I've tested this and it works as described. 

But I'd like to clarify the following scenario. 
- Customer installs and activate a product.
- Later the customer will delete the plugin/extension in the local store.
- Then they will try to install it again. (I assume the new installation will use a new `idempotency-key`)
- But the installation will fail due to miss match in `idempotency-key`
```
{"success":false,"error_code":"idempotency_key_mismatch","error_message":"Idempotency key mismatch","state":{"product_id":390890,"idempotency_key":"ABC","last_step_name":"activate_product","last_step_status":"completed","last_step_error":null,"product_type":"plugin","product_name":"WooCommerce Bookings","already_installed_plugin_info":null,"started_seconds_ago":307}
```

If the client of the API is not keeping track of the `idempotency-key` they won't even be able to reset the installation.

**Comment by rcstr** [comment]
I noticed that after I reset the state, `already_installed_plugin_info` had the information. I guess is because the `wp_option` is created; in this case, the plugin was already downloaded from the previous installation.

### reset request
```sh
$ curl --request DELETE \
  --url 'http://localhost:8888/wp-json/wccom-site/v2/installer?token=XXX&signature=YYY' \
  --header 'Content-Type: application/json' \
  --data '{
        "product-id": 5278104,
        "idempotency-key": "ABC"
}'

{"success":true,"state":null}
```

### Repeat reset request
```sh
$ curl --request DELETE   --url 'http://localhost:8888/wp-json/wccom-site/v2/installer?token=XXX&signature=YYY'   --header 'Content-Type: application/json'   --data '{
        "product-id": 5278104,
        "idempotency-key": "ABC"
}'

{
	"success": false,
	"error_code": "no_initiated_installation_found",
	"error_message": "No initiated installation for the product found",
	"state": null
}

```

### Install again
```sh
$ curl --request POST   --url 'http://localhost:8888/wp-json/wccom-site/v2/installer?token=XXX&signature=YYY'   --header 'Content-Type: application/json'   --data '{
        "product-id": 5278104,
        "run-until-step": "activate_product",
        "idempotency-key": "ABC"
}'

{
	"success": true,
	"state": {
		"product_id": 5278104,
		"idempotency_key": "ABC",
		"last_step_name": "activate_product",
		"last_step_status": "completed",
		"last_step_error": null,
		"product_type": "plugin",
		"product_name": "WooCommerce Payments",
		"already_installed_plugin_info": {
			"name": "WooCommerce Payments",
			"version": "5.7.0",
			"active": true
		},
		"started_seconds_ago": 3
	}
}

```

**Review by anaspk — DISMISSED** [review]
Thank you @denho , LGTM and works as described.

I can confirm that I have also observed the same behaviors as indicated by Rommel and Thilinah. I think these might not necessarily be wrong/buggy behaviors, so I am going to approve it. Feel free to make any amendments as you see fit.

**Review by thilinah — DISMISSED** [review]
@denho changes look good. I've added a [comment](https://github.com/woocommerce/woocommerce/pull/37753#issuecomment-1513505703) but it's not a blocker since the API client can use the key from failed response to reset the installation if needed.

**Review by rcstr — DISMISSED** [review]
LGTM! thanks for working on this refactor, the code is much better and will be much easier to work with

**Comment by denho** [comment]
> @denho while testing I noticed that already_installed_plugin_info is null even after trying to reinstall a few times. I tried with a few products and had the same result

Thanks for raising this, @rcstr. Actually, that behavior is expected. When you repeat the installation request, after the previous one succeeded, a new installation will not be initiated - it will return an error instead, as you have provided:
```
        "success": false,
	"error_code": "all_installation_steps_run",
	"error_message": "All installation steps have been run",
...
```
In that case, as a new installation hadn't been initiated, it will return the state unchanged from the previous installation, which was successful, and for that installation the `already_installed_plugin_info` is expectedly `null`:
```
"state": {
		"product_id": 5278104,
		"idempotency_key": "ABC",
		"last_step_name": "activate_product",
		"last_step_status": "completed",
		"last_step_error": null,
		"product_type": "plugin",
		"product_name": "WooCommerce Payments",
		"already_installed_plugin_info": null,
		"started_seconds_ago": 31
	}
```

But, as you mentioned further, after resetting the state, which thus allows to trigger a new installation, it will return the `already_installed_plugin_info` with data, which reflects that the plugin has been already installed when the installation was triggered.

> If the client of the API is not keeping track of the idempotency-key they won't even be able to reset the installation.

@thilinah The recent `idempotency-key` is always returned in the state, even in the failed requests due to idempotency key mismatch:
```
{
		"success":false,
		"error_code":"idempotency_key_mismatch",
		"error_message":"Idempotency key mismatch",
		"state":
		{
				"product_id":390890,
				"idempotency_key":"ABC",
		...
```
This allows to get that recent idempotency key from the response and use it in the reset state request, which unblock the new installation.

File: plugins/woocommerce/includes/wccom-site/rest-api/class-wc-rest-wccom-site-installer-error-codes.php [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+	const REQUESTED_STEP_ALREADY_RUN          = 'requested_step_already_run';
+	const PLUGIN_ALREADY_INSTALLED            = 'plugin_already_installed';
+	const INSTALLATION_ALREADY_RUNNING        = 'installation_already_running';
+	const INSTALLATION_FAILED                 = 'installation_failed';
+	const FILESYSTEM_REQUIREMENTS_NOT_MET     = 'filesystem_requirements_not_met';
+	const FAILED_GETTING_PRODUCT_INFO         = 'product_info_failed';
+	const INVALID_PRODUCT_INFO_RESPONSE       = 'invalid_product_info_response';
+	const WCCOM_PRODUCT_MISSING_SUBSCRIPTION  = 'wccom_product_missing_subscription';
+	const WCCOM_PRODUCT_MISSING_PACKAGE       = 'wccom_product_missing_package';
+	const WPORG_PRODUCT_MISSING_DOWNLOAD_LINK = 'wporg_product_missing_download_link';
```
Comment (anaspk) [inline plugins/woocommerce/includes/wccom-site/rest-api/class-wc-rest-wccom-site-installer-error-codes.php:38]: @denho I just noticed that this `WPORG_PRODUCT_MISSING_DOWNLOAD_LINK` code is missing in the `ERROR_MESSAGES` array below. Is that intentional?

Comment (denho) [reply plugins/woocommerce/includes/wccom-site/rest-api/class-wc-rest-wccom-site-installer-error-codes.php:38]: Not intentional, but that's fine. It was originally designed that messages and HTTP codes are optional, and there is a fallback to default value:
https://github.com/woocommerce/woocommerce/blob/56b63ada60e209003fb69524c76c7e8630392b0a/plugins/woocommerce/includes/wccom-site/rest-api/class-wc-rest-wccom-site-installer-error.php#L25-L26

**Review by jorgeatorres — APPROVED** [review]
LGTM. Thank you @denho!

**Pull request merged**